File: src/main/java/com/jonas/test/functional/nullable/Tester.java
package com.jonas.test.functional.nullable;

import java.util.function.Supplier;

/**
 * 测试在传递 lambada 引用时，源实例是否可以为 null
 * Created by jonas on 2017/3/31.
 */
public class Tester {
    public static void main(String[] args) {
        Model model = null;
        System.out.println(getIfNotNull(model, model::getName));
    }

    public static <T> String getIfNotNull(T object, Supplier<String> supplier) {
        return object != null ? supplier.get() : null;
    }
}

class Model {
    String getName() {
        return "World";
    }
}

File: src/main/java/com/jonas/test/json/jackson/dates/Tester.kt
package com.jonas.test.json.jackson.dates

import com.fasterxml.jackson.annotation.JsonFormat
import com.fasterxml.jackson.databind.ObjectMapper
import java.sql.Timestamp
import java.time.Instant
import java.time.LocalDateTime
import java.util.*

/**
 * 测试 Jackson 的时间序列化
 * Created by jonas on 2017/3/31.
 */

fun main1(args: Array<String>) {
    var model = Model2()
    val timestamp = LocalDateTime.now()
    model.created = Timestamp.valueOf(timestamp)
    val objectMapper = ObjectMapper()
    val json = objectMapper.writeValueAsString(model)
    println(json)
    println(timestamp)
    model = objectMapper.readValue(json, Model2::class.java)
    println(model.created)
}

fun main(args: Array<String>) {
    val json = """{"created":"2017-03-31 14:54:51"}"""
    val objectMapper = ObjectMapper()
    val model = objectMapper.readValue(json, Model1::class.java)
    println(model.created)

//    for (zone in TimeZone.getAvailableIDs(0)) {
//        println(zone)
//    }
}

class Model1 {

    @JsonFormat(shape= JsonFormat.Shape.STRING, pattern="yyyy-MM-dd hh:mm:ss", timezone = "UTC")
    var created: Date? = null;
}

class Model2 {
    var created: Timestamp? = null;
}

class Model3 {
    var created: Instant? = null;
}